lib/src/visitor/serialize.dart

```diff
@@ -1078,43 +1080,59 @@ class _SerializeVisitor
       _for(value, () => _buffer.write(value.value));
 
   /// Emits [children] in a block.
-  void _visitChildren(List<CssNode> children) {
+  void _visitChildren(List<CssNode> children, CssParentNode parent) {
     _buffer.writeCharCode($lbrace);
-    if (children.every(_isInvisible)) {
-      _buffer.writeCharCode($rbrace);
-      return;
-    }
 
-    _writeLineFeed();
-    CssNode previous;
-    _indent(() {
-      for (var i = 0; i < children.length; i++) {
-        var child = children[i];
-        if (_isInvisible(child)) continue;
+    CssNode previous = parent;
+    for (var child in children) {
+      if (_isInvisible(child)) continue;
+      if (previous != parent) {
+        if (_requiresSemicolon(previous)) _buffer.writeCharCode($semicolon);
+      }
+      if (_isTrailingComment(child, previous)) {
+        _writeOptionalSpace();
+        _withoutIndendation(() => child.accept(this));
+      } else {
+        _writeLineFeed();
+        _indent(() {
+          child.accept(this);
+        });
+      }
 
-        if (previous != null) {
-          if (_requiresSemicolon(previous)) _buffer.writeCharCode($semicolon);
-          _writeLineFeed();
-          if (previous.isGroupEnd) _writeLineFeed();
-        }
-        previous = child;
+      previous = child;
+    }
 
-        child.accept(this);
+    if (previous != parent) {
+      if (_requiresSemicolon(previous) && !_isCompressed) {
+        _buffer.writeCharCode($semicolon);
       }
-    });
-
-    if (_requiresSemicolon(previous) && !_isCompressed) {
-      _buffer.writeCharCode($semicolon);
+      _writeLineFeed();
+      _writeIndentation();
     }
-    _writeLineFeed();
-    _writeIndentation();
+
     _buffer.writeCharCode($rbrace);
   }
 
   /// Whether [node] requires a semicolon to be written after it.
   bool _requiresSemicolon(CssNode node) =>
       node is CssParentNode ? node.isChildless : node is! CssComment;
 
+  /// Whether [node] represents a trailing comment when it appears after
+  /// [previous] in a sequence of nodes being serialized.  Note [previous]
+  /// could either be (1) a sibling of [node] or (2) the parent of [node], with
+  /// [node] being the first visible child.
+  bool _isTrailingComment(CssNode node, CssNode previous) {
+    if (node is! CssComment) return false;
+    FileSpan previousSpan = previous.span;
+    if (previous is CssParentNode && previous.children.contains(node)) {
+      var lbracePattern = String.fromCharCode($lbrace);
+      var endOffset = math.max(0, previousSpan.text.indexOf(lbracePattern));
```

Comment:
It's an edge case, but this `indexOf()` check could fail if a selector or at-rule contains curly braces (for example, `[href*="{"]` is a valid selector), or even if it contains a comment that was removed from the AST (such as `a /*{*/ b { ... }`). It would be safer to look backwards from `node.span.start`.

Reply:
I added a test case for this, confirmed it failed, and then fixed it (by doing what you suggested).  I'll push those as soon as all the other tests finish running.

Reply:
committed in https://github.com/nickbehrens/dart-sass/commit/45afc0a31b5115487fa816120aff32019890d244